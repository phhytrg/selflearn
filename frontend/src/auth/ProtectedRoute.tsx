import { Link, Navigate, Outlet } from 'react-router-dom';
import { useAuth } from './hooks/useAuth';
import { Button, Drawer } from 'antd';
import { useState } from 'react';
import { MenuOutlined } from '@ant-design/icons';
import { useRoles } from '@/shared/hooks/useAuthority';

export const ProtectedRoute = () => {
  const { isAdmin } = useRoles();
  const [open, setOpen] = useState(false);
  const { user, logout } = useAuth();

  const showDrawer = () => {
    setOpen(true);
  };

  const onClose = () => {
    setOpen(false);
  };

  if (!user) {
    //Try to get user from local storage
    return <Navigate to="/login" />;
  }

  return (
    <div className="flex flex-col">
      <Button
        className="ml-auto"
        aria-label="menu"
        icon={<MenuOutlined />}
        onClick={showDrawer}
      />
      <Drawer
        data-testid="drawer"
        onClose={onClose}
        open={open}
        styles={{
          body: {
            display: 'flex',
            flexDirection: 'column',
            gap: '8px',
          },
        }}
      >
        <Link to={'/Home'}>Home</Link>
        {!isAdmin || <Link to={'/Admin'}>Admin</Link>}
        <a onClick={logout}>Log out</a>
      </Drawer>
      <Outlet />
    </div>
  );
};
