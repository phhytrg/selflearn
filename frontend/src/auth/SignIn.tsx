import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, Input, Modal, Spin } from 'antd';
import { useAuth } from './hooks/useAuth';

const LoginPage = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate();
  const { login, user } = useAuth();
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (user) {
      navigate('/', { replace: true });
    }
  }, [user, navigate]);

  const handleLogin = async (e: any) => {
    e.preventDefault();
    try {
      setIsLoading(true);
      const res = await login({
        username: email,
        password,
      });
      console.log(res);
    } catch (e: any) {
      Modal.error({
        title: 'Error',
        content: (
          <div data-testid="error-modal">
            {JSON.stringify(e.response?.data) || e.message}
          </div>
        ),
      });
      return;
    } finally {
      setIsLoading(false);
    }
    navigate('/', { replace: true });
  };

  return (
    <div className="flex flex-col gap-4">
      <div>
        <label htmlFor="email">Email:</label>
        <Input
          id="email"
          type="text"
          name="email"
          aria-label="email"
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
      </div>
      <div>
        <label htmlFor="password">Password:</label>
        <Input
          id="password"
          type="password"
          name="password"
          aria-label="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <Button
        type="primary"
        onClick={handleLogin}
        name="signIn"
        aria-label="signIn"
        disabled={isLoading}
      >
        {isLoading ? <Spin /> : `Log In`}
      </Button>
      <Button
        type="default"
        name="signUp"
        aria-label="signUp"
        onClick={() => {
          navigate('/signup', { replace: true });
        }}
      >
        Sign Up
      </Button>
    </div>
  );
};

export default LoginPage;
