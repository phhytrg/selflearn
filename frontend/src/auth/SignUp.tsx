import { Button, Input, Modal, Spin } from 'antd';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useAuth } from './hooks/useAuth';
import { authApi } from '@/shared/apis';

export const SignUpPage = () => {
  const { user } = useAuth();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const navigate = useNavigate();
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (user) {
      navigate('/', { replace: true });
    }
  }, [user, navigate]);

  const handleSignUp = async (e: any) => {
    e.preventDefault();
    setIsLoading(true);
    try {
      const res = await authApi.register(email, password, confirmPassword);
    } catch (e: any) {
      Modal.error({
        title: 'Error',
        content: (
          <p data-testid="error-modal">
            {Object.entries(e.response.data).map(([k, v]: any) => {
              // Explicitly type 'v' as string
              return (
                <p key={k}>
                  {k}: {v}
                </p>
              );
            })}
          </p>
        ),
      });
      return;
    }
    setIsLoading(false);
    navigate('/login', { replace: true });
  };

  return (
    <div className="flex flex-col gap-4">
      <h1>Sign Up</h1>
      <Input
        type="text"
        placeholder="Email"
        value={email}
        aria-label="email"
        onChange={(e) => {
          setEmail(e.target.value);
        }}
      />
      <Input
        type="password"
        aria-label="password"
        placeholder="Password"
        value={password}
        onChange={(e) => {
          setPassword(e.target.value);
        }}
      />
      <Input
        type="password"
        placeholder="Confirm Password"
        aria-label="confirmPassword"
        value={confirmPassword}
        onChange={(e) => {
          setConfirmPassword(e.target.value);
        }}
      />
      <Button
        type="primary"
        onClick={handleSignUp}
        aria-label="signUp"
        disabled={isLoading}
      >
        {isLoading ? <Spin /> : `Sign Up`}
      </Button>
    </div>
  );
};
