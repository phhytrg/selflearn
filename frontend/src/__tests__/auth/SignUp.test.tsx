import routesConfig from '@/routes';
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from 'vitest';
import { server } from '../msw/worker';
import { handlers as authHandlers } from '../msw/mocks/auth.handlers';
import { faker } from '@faker-js/faker';

describe('SignUp Component', () => {
  vi.mock('antd', async (importOriginal) => {
    const mod = await importOriginal<typeof import('antd')>();
    return {
      ...mod,
      Spin: () => {
        return <div aria-label="spin">Spin</div>;
      },
    };
  });

  beforeAll(() => {
    server.use(...authHandlers);
  });

  beforeEach(() => {
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ['/signup'],
    });
    render(<RouterProvider router={router} />);
  });

  afterEach(() => {
    cleanup();
  });

  it('should render SignUp component', () => {
    expect(screen.getByPlaceholderText('Email')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Confirm Password')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'signUp' })).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    cleanup();
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ['/signup'],
    });
    const { asFragment } = render(<RouterProvider router={router} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should navigate to login page when user sign up successfully', async () => {
    const signUpButton = screen.getByRole('button', { name: 'signUp' });
    const emailInput = screen.getByPlaceholderText('Email');
    const passwordInput = screen.getByPlaceholderText('Password');
    const confirmPasswordInput =
      screen.getByPlaceholderText('Confirm Password');

    fireEvent.change(emailInput, { target: { value: 'test1@gmail.com' } });
    fireEvent.change(passwordInput, { target: { value: 'password' } });
    fireEvent.change(confirmPasswordInput, { target: { value: 'password' } });
    fireEvent.click(signUpButton);

    await waitForElementToBeRemoved(() => {
      const spinElement = screen.getByLabelText('spin');
      waitFor(() => expect(spinElement).not.toBeInTheDocument());

      return spinElement;
    });

    expect(screen.getByRole('button', { name: 'signIn' })).toBeInTheDocument();
  });

  it('should show error modal when password and confirm password do not match', async () => {
    const signUpButton = screen.getByRole('button', { name: 'signUp' });
    const emailInput = screen.getByPlaceholderText('Email');
    const passwordInput = screen.getByPlaceholderText('Password');
    const confirmPasswordInput =
      screen.getByPlaceholderText('Confirm Password');

    fireEvent.change(emailInput, { target: { value: faker.internet.email() } });
    fireEvent.change(passwordInput, { target: { value: 'password' } });
    fireEvent.change(confirmPasswordInput, { target: { value: 'password1' } });
    fireEvent.click(signUpButton);

    await waitFor(() => {
      expect(screen.getByTestId('error-modal')).toBeInTheDocument();
    });
  });

  afterAll(() => {
    console.log('msw handlers: ', server.listHandlers());
    // server.resetHandlers();
  });
});
