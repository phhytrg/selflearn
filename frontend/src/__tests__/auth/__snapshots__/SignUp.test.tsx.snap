// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUp Component > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4"
  >
    <h1>
      Sign Up
    </h1>
    <input
      aria-label="email"
      class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      aria-label="password"
      class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      aria-label="confirmPassword"
      class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
      placeholder="Confirm Password"
      type="password"
      value=""
    />
    <button
      aria-label="signUp"
      class="ant-btn css-dev-only-do-not-override-kghr11 ant-btn-primary"
      type="button"
    >
      <span>
        Sign Up
      </span>
    </button>
  </div>
</DocumentFragment>
`;
